src/examples/libpmem2/redo.c

```diff
@@ -0,0 +1,388 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/* Copyright 2020, Intel Corporation */
+
+/*
+ * basic.c -- simple example for the libpmem2
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+#ifndef _WIN32
+#include <unistd.h>
+#else
+#include <io.h>
+#endif
+#include <libpmem2.h>
+
+#define REDO_LOG_SIZE 1000
+#define MINIMUM_POOL_SIZE (sizeof(struct pool_hdr) + sizeof(struct node) * 100)
+#define offset(pool, addr) ((uintptr_t)addr - (uintptr_t)pool)
+
+static pmem2_persist_fn Persist;
+static pmem2_flush_fn Flush;
+static pmem2_drain_fn Drain;
+static pmem2_memcpy_fn Memcpy;
+
+struct redo_log_entry {
+	uintptr_t offset;
+	uint64_t data;
+};
+
+struct redo_log {
+	uint64_t last;
+	uint8_t apply;
+	uint8_t unused[57]; /* align to cache line */
+	struct redo_log_entry entries[REDO_LOG_SIZE];
+};
+
+struct node {
+	uint64_t this;
+	uint64_t prev;
+	uint64_t next;
+	uint64_t key;
+	uint64_t value;
+};
+
+struct pool_hdr {
+	struct redo_log redo;
+	uint64_t free;
+	uint64_t head;
+	uint64_t pool_size;
+};
+
+struct pool_layout {
+	struct pool_hdr hdr;
+	struct node nodes[];
+};
+
+/*
+ * redo_apply -- process and apply the redo log
+ */
+static void
+redo_apply(struct pool_layout *pool)
+{
+	struct redo_log *redo = &pool->hdr.redo;
+
+	if (!redo->apply || !redo->last) {
+		if (redo->apply || redo->last) {
+			redo->last = 0;
+			redo->apply = 0;
+			Persist(&redo, 256); /* persist entire CL */
+		}
+		return; /* no redo log to process */
+	}
+
+	uint8_t *start = (uint8_t *)pool;
+	for (int i = 0; i < redo->last; ++i) {
+		uint64_t *node = (uint64_t *)&start[redo->entries[i].offset];
```

Comment:
I'd prefer `start + offset;`, this might make someone think that `start` is an actual array.